### PR TITLE
Sample operator

### PR DESCRIPTION
**Pull request by MarekLabuz**
Hi, I've implemented Sample operator + tests. I've chosen the second variant described in the documentation (that is Observable as an argument), because it is more general and the operator can be used in a broader range.

### CODE REVIEW

File: observable.go [hunk 1]
```diff
@@ -1111,6 +1112,69 @@ func (o *observable) Repeat(count int64, frequency Duration) Observable {
...
+					isAnyItemEmitted = true
+				} else {
+					cancel()
+					close(out)
+					break mainLoop
+				}
+			case _, ok := <-obsChan:
+				if ok {
+					if isAnyItemEmitted {
+						out <- lastEmittedItem
```
Comment (teivah) [inline observable.go]: Shouldn't we put `lastEmittedItem` to true then?

According to the documentation, the following code should print only one _1_:
```go
ch := make(chan interface{}, 1)
obs := FromChannel(ch)
sample := obs.Sample(Interval(make(chan struct{}), time.Second))
sample.Subscribe(handlers.NextFunc(func(i interface{}) {
	fmt.Printf("%v\n", i)
}))
ch <- 1
```

Here, though, it keeps printing 1.

**Comment by teivah** [comment]
> because it is more general and the operator can be used in a broader range

That's a good callout in my opinion as well.

File: observable.go (hunk 1 above)
Comment (MarekLabuz) [reply observable.go]: You're right. I have not tested such situation. I've just pushed changes that include fixed solution and tests.

File: observable.go [hunk 2]
```diff
@@ -1111,6 +1112,70 @@ func (o *observable) Repeat(count int64, frequency Duration) Observable {
...
+				if item, err := it.Next(ctx); err == nil {
+					obsChan <- item
+				} else {
+					break
+				}
+			}
+			close(obsChan)
+		}()
+
+	mainLoop:
```
Comment (teivah) [inline observable.go]: One very last comment and we can merge it. Can we get rid of this label and replace the two `break mainLoop` by `return`. I prefer avoiding label unless it is absolutely mandatory.

**Comment by teivah** [comment]
@MarekLabuz What's your opinion on these tests using ginkgo & gomega? Do you find this useful and easy to implement?

**Comment by MarekLabuz** [comment]
I've just pushed the changes. Regarding ginkgo & gomega, yes, I'm satisfied with them, because tests are well structured. However, I see that you are trying to get rid of them - is it due to the performance and testing time?

**Comment by teivah** [comment]
@MarekLabuz Man, you're going to hate me!
Well, after my comment on the tests, I spent my evening in refactoring the tests. It dropped the execution time from 16 seconds to less than 500 milliseconds. This explains the conflict 😊.
There was no good reason to not leverage our internal testing framework.

**Comment by MarekLabuz** [comment]
Not a problem ;) I'll look into the changes you've made and fix my branch

**Comment by MarekLabuz** [comment]
@teivah Okay, it should be alright now. Please check whether I did it correctly :)

**Comment by teivah** [comment]
I've been thinking a lot about testing recently. I feel like if we are obliged to use sleeps in our test it means something is broken. But obviously, for this `Sample` operator it was not easy to use something else.

I tried to implement this (if you want to take a look and give me your opinion that would be awesome): https://github.com/ReactiveX/RxGo/blob/mock/observable_test.go#L1645

Basically, there's a DSL to create iterators. For example:
```
1
2
	0
x
	x
```

It means the iterator one will send `1`, `2`. Then, the iterator 2 (the _timer_) will send a `0` (just to notify a tick), then the iterator one will close (`x` value) and so the iterator 2. Each column corresponds to one iterator.

The huge difference is that there is a causality between each item. The `9` item cannot be sent until the `2` item has been sent and handled by the observable. This lead to deterministic tests (i.e. no sleep).

I'm not sure how to explain better, it's kinda late ^^

**Pull request merged**